### ISSUE TITLE
Crash when object is freed while it is waiting for server reply

### ISSUE THREAD

**Issue by kb173** [boku-ilen/landscapelab#15]
It can currently happen frequently that a WorldTile is deleted before it gets all its data from the server, e.g. if moving to the right, spawning new tiles, and then moving back to the left, deleting those tiles again.

Because the function which waits for a reply from the server is run in a thread, the tile can be deleted while this process is still running. Thus, once the reply comes, the node is actually already freed, causing a crash when trying to change its values.

Constantly checking if the node is still valid seems like a hacky solution. Unfortunately however, threads can't just be killed. Should we postpone all free()-ing until all threads are done by writing a wrapper around it?

**Comment by chrisgraf** [comment]
Is it simply possible to check for the existence of the node before making a change?

**Comment by kb173** [comment]
This and #10 was much improved by commit d8b7280. It's not completely clear why, but many of the crashes were caused by WorldTiles getting `free`d. In the case described in this issue, it makes sense, but even when waiting for all threads to finish, crashes occured... For this reason, WorldTiles are currently never deleted, only made invisible, which doesn't seem to have much of a performance impact.

There are still occasional crashes, but they're now limited to C++ engine errors such as:
`malloc(): smallbin double linked list corrupted`
`malloc_consolidate(): invalid chunk size`
`malloc(): memory corruption (fast)`
`double free or corruption (out) \n INFO: code: 200`
`handle_crash: Program crashed with signal 11`

It seems like sometimes, nodes are being deleted even though there is no reason to, even engine Singletons, like in this example where the `logger` seems to be deleted:

```
SCRIPT ERROR: try_to_get_json: Attempt to call function 'info' in base 'previously freed instance' on a null instance.
   At: res://Global/ServerConnection.gd:93.
ERROR: get_packet: Condition ' remaining < 4 ' is true. returned: ERR_UNAVAILABLE
   At: core/io/packet_peer.cpp:220.
ERROR: _poll_events: Condition ' err != OK ' is true. Continuing..:
   At: core/script_debugger_remote.cpp:691.
```

Rarely, I get a backtrace, although it's not too helpful either:

```
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x3ef20) [0x7f527e944f20] (??:0)
[2] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0x24bfe79] (??:?)
[3] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0xb8cd2a] (<artificial>:?)
[4] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0xbe1af6] (<artificial>:?)
[5] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0xb772c0] (<artificial>:?)
[6] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0xb767ec] (<artificial>:?)
[7] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0xbd20b9] (<artificial>:?)
[8] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0xb424e5] (<artificial>:?)
[9] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0x917af8] (<artificial>:?)
[10] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0xaf30b2] (??:?)
[11] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0x7b038d] (??:?)
[12] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f527e927b97] (??:0)
[13] /home/karl/Downloads/Godot_v3.1-beta10_x11.64() [0x7bddbe] (??:?)
-- END OF BACKTRACE --
```

**Comment by kb173** [comment]
Another note: When "TerrainModule.tscn" is the only activated module, I wasn't able to get the client to crash at all. Same when also spawning "TerrainColliderModule.tscn", but not creating the collider shape. It seems like there might be a tipping point in workload which makes strange things happen in the engine...

**Comment by kb173** [comment]
The reason why only the higher LODs are reliably causing crashes seems to be largely the `get_cropped` function in `CachingImageTexture.gd`, which is mostly called at those LODs (since there, the server can't provide all textures anymore) - not using it makes the client almost completely stable. Will investigate this further

**Comment by chrisgraf** [comment]
Does this happen on windows too?

**Comment by kb173** [comment]
Last time I checked, it did; once I have my development environment set up on my Windows machine again, I will try again to see if anything has changed there.

**Comment by kb173** [comment]
It seems like the only way to fix this in a completely general way would require modifying the engine. For this reason, https://github.com/boku-ilen/landscapelab/commit/695acdecd11a21cc030ee13a447867ffd8e273aa fixes this for WorldTiles, which are the only node that can be affected by this issue.

Essentially, a `delete()` function has been added, which wraps the engine's `queue_free()`. This function simply sets a flag which makes the node delete itself only once all its modules are done loading (meaning their Threads are all finished) and all its children have successfully deleted themselves as well.
